README.md
canceR
======

An R package housing dynamical system models of non-myeloid cancers



canceR/NEWS
This package contains models of nonmyeloid cancers.

Version 0.1.0   12/10/2012: First Draft containing hahnfeldt99 
                    
                   


canceR/R/hahnfeldt99.R
#'Antiangiogenesis model of Hahnfeldt et al. Cancer Research 1999
#'
#'This function returns the right hand side of an ordinary differential
#'equation model of antiangiogenesis and tumor growth that was published by Hahnfeldt et al
#'in Cancer Research in 1999.  The intended use of this
#'function is as an argument to \code{ode()} of the \code{deSolve} package.
#'
#'The model captures tumor carrying capacity dynamics. 
#'
#'@param Time The parameters of this model do not depend on time: this argument
#'exists here as a dummy only because deSolve expects it to exist.
#'@param State Vector of current states. The elements of this vector must be
#'named because within this function's definition is
#'\code{with(as.list(State,Pars), {XXX} )} and code in XXX presupposes that it
#'can call named elements as variables with those names.
#'@param Pars Vector of parameter values with \emph{named} (see above)
#'elements.
#'@return A list of length 2 (as expected by deSolve) where the first list
#'element is the vector of derivatives, i.e. the right hand side of the ODEs 
#'and the second element of the list is a vector of auxiliary
#'variables that one may wish to track over time.
#'@note This work was supported by the National Cancer Institute and Tufts
#'Integrative Cancer Biology Program under U54CA149233-029689.
#'@author Tom Radivoyevitch (\email{txr24@@case.edu})
#'@seealso \code{\link{canceR-package}}
#'@references Philip Hahnfeldt, Dipak Panigrahy, Judah Folkman, and Lynn Hlatky, Tumor Development under Angiogenic 
#'Signaling: A Dynamical Theory of Tumor Growth, Treatment Response, and Postvascular Dormancy,
#'\emph{Cancer Research},
#'\bold{59}, 4770-4775 (1999)
#'@keywords IO
#'@export
#'@examples
#'\dontrun{
#'library(canceR)
#'out1=lsoda(y=c(V=180,K= 625,g=0),times=seq(0,20,1),hahnfeldt99,  
#' 	parms=c(lambda1=.192,lambda2=0,b=5.85,d=0.00873,e=0,clr=0),rtol=1e-4, atol= rep(1e-4,3)) 
#'plot(out1) 
#'}

hahnfeldt99<-function(Time, State, Pars) {
	with(as.list(c(State, Pars)), {
				dV = -lambda1*V*log(V/K)
				dK = -lambda2*K+b*V-d*K*V^.667-e*K*g
				dg = -clr*g
				return(list(c(dV,dK,dg)))
			})
}


canceR/demo/hahnfeldtDemo.R
# first growth with no treatment, from help page
library(canceR)
out1=lsoda(y=c(V=180,K= 625,g=0),times=seq(0,20,1),hahnfeldt99,  
		parms=c(lambda1=.192,lambda2=0,b=5.85,d=0.00873,e=0,clr=0),rtol=1e-4, atol= rep(1e-4,3)) 
plot(out1) 

# now do Angiostatin and Endostatin together, daily for 24 days
hahnfeldt99AE<-function(Time, State, Pars) {
	with(as.list(c(State, Pars)), {
				dV = -lambda1*V*log(V/K)
				dK = -lambda2*K+b*V-d*K*V^.667-eA*K*gA-eE*K*gE
				dgA = -clrA*gA
				dgE = -clrE*gE
				return(list(c(dV,dK,dgA,dgE)))
			})
}

library(deSolve)
y0<-c(V=180,K= 625,gA=20,gE=20) # angio and endo high dose combined
outs=NULL
eventdat <- data.frame(var = rep(c("gA", "gE"),25),
		time = rep(0:24,each=2) ,
		value = rep(20,50),
		method = c("add"))
eventdat

out=lsoda(y=y0,times=seq(0,30,.2),hahnfeldt99AE, parms=c(lambda1=.192,lambda2=0,b=5.85,d=0.00873,
					eA=0.15,clrA=0.38,eE=0.66,clrE=1.7), rtol=1e-4, atol= rep(1e-4,4),events = list(data = eventdat))
plot(out)


out1=lsoda(y=c(V=180,K= 625,g=0),times=seq(0,20,1),hahnfeldt99,  
		parms=c(lambda1=.192,lambda2=0,b=5.85,d=0.00873,e=0,clr=0),rtol=1e-4, atol= rep(1e-4,3)) 
plot(out1) 

# now check to see if DLL's are woring the same way
(f=file.path(system.file(paste("libs",Sys.getenv("R_ARCH"),sep=""), package = "canceR"),
					paste("canceR",.Platform$dynlib.ext,sep="")))
dyn.load(f)
#dlls <- getLoadedDLLs()
#getDLLRegisteredRoutines(dlls[["canceR"]])
#getDLLRegisteredRoutines("canceR")
#getNativeSymbolInfo("canceR")

out1=ode(y=c(V=180,K= 625,g=0),times=seq(0,20,1),  
		parms=c(lambda1=.192,lambda2=0,b=5.85,d=0.00873,e=0,clr=0),rtol=1e-4, atol= rep(1e-4,3),
		func = "derivsPhil", dllname = "canceR",initfunc = "initPhil" )
plot(out1) 

out=lsoda(y=y0,times=seq(0,30,.2),parms=c(lambda1=.192,lambda2=0,b=5.85,d=0.00873,
eA=0.15,clrA=0.38,eE=0.66,clrE=1.7), rtol=1e-4, atol= rep(1e-4,4),events = list(data = eventdat),
				func = "derivsPhilAE", dllname = "canceR",initfunc = "initPhilAE" )
plot(out)

######################################################################################
# Warning: the following code is old to the extent that it uses neither bbmle nor events
#########################################################################################
(fsrc=file.path(system.file("demo/hahnfeldtDemoFuncs.R", package = "canceR")))
source(fsrc)

g=NULL
g$dC=data.frame( days=c(0,4,7,10,13,16,19), size=c(192,728,2319,3590,5852,8591,10438))
g$dT=data.frame( days=c(0,4,7,10,13 ), size=c(170,381,875,1363,1750 ))
g$dA=data.frame( days=c(0,4,7,10,13), size=c(176,246,265,136,207))
g$dE=data.frame( days=c(0,4,7,10 ), size=c(186,157,135,76 ))
g$dEL=data.frame( days=c(0,4,7,10,13,16), size=c(291,429,859,2420,3120,4100)) # endo at 4mg/kg
g$dAE=data.frame( days=c(0,4,7,10,13,16,19,22,25), size=c(309,180,88,79,29,6,0,0,0))
g$nData=dim(g$dC)[1]+dim(g$dT)[1]+dim(g$dA)[1]+dim(g$dE)[1]+dim(g$dEL)[1]+dim(g$dAE)[1]
parICs=c(lambda1=.192,lambda2=0,b=5.85,d=0.00873,eT=1.3,clrT=10.1,eA=0.15,clrA=0.38,eE=0.66,clrE=1.7,V0=180,V0valid=300,K0=625)
g$params=data.frame(initial=parICs,final=parICs,opt=TRUE,CI95prct="not fitted")
g


g=simData(g)
# first take a look at the initial parameter value situation
plotg(g)  # mg/kg time courses
plotVK(g) # Tumor volume setpoint and process variable time courses
g$params

g$params[c("lambda2"),"opt"]=FALSE   
#g$params[c("lambda2","eT","clrT","eA","clrA","eE","clrE"),"opt"]=FALSE  # only fit control parameters and V0valid
p0=t(g$params[g$params[,"opt"],"initial",drop=F])
p0=unlist(as.data.frame(p0))
p0=sapply(p0,log)
g$optNames<-names(p0)
g

t0 = Sys.time();  
opt<-optim(p0,fopt,hessian=TRUE,model=g,control=list(trace=F,maxit=5000))
difftime(Sys.time(),t0,units="mins")[[1]]

opar<-opt$par
Hess=opt$hessian
value=opt$value
sg<-sqrt(value/(g$nData-length(p0)))
if (abs(det(Hess))>1e-8) {
	sig=sg*sqrt(diag(solve(Hess/2)))
	upper=signif(opar+1.96*sig,3)
	lower=signif(opar-1.96*sig,3)
	g$CI=exp(cbind(lower,upper))
}  
g$params[g$optNames,"final"]=signif(exp(opar),3)  
g$params[g$optNames,"CI95prct"]=paste("(",signif(g$CI[,"lower"],3),", ",signif(g$CI[,"upper"],3),")",sep="")  
g$params

# now look at final parameter estimate situation
g=simData(g)
plotg(g)  # mg/kg time courses
plotVK(g) # Tumor volume setpoint and process variable time courses
dyn.unload(f)


canceR/demo/hahnfeldtDemoFuncs.R
updateSSE<-function(dX,oldSSE) {
#  oldSSE+sum((dX$EV-dX$size)^2)  #too much emphasis on control data
	oldSSE+sum((sqrt(dX$EV)-sqrt(dX$size))^2)  # sqrt transform: still not great since K0 still less than V0
#  oldSSE+sum((dX$EV-dX$size)^2)/mean(dX$size)^2 # normalize: this led to K0 that are too small to believe
#  oldSSE+sum((log(dX$EV)-log(dX$size))^2) # problem is that some volumes are zero
}

simData<-function(g){
	SSE=0
	Pframe=as.data.frame(t(g$params[,"final",drop=F]))
	attach(Pframe)
	y0<-c(V=V0,K= K0,g=0)
	out1=ode(y=y0,times=seq(0,20,1),parms=c(lambda1=lambda1,lambda2=0,b=b,d=d,e=0,clr=0), 
			rtol=1e-4, atol= rep(1e-4,3),
			func = "derivsPhil", dllname = "canceR",initfunc = "initPhil" )

	g$outC=as.data.frame(out1)
	g$dC$EV=g$outC[match(g$dC$days,g$outC$time),"V"]
	SSE=updateSSE(g$dC,0)
	
	y0<-c(V=V0,K= K0,g=30)
	outs=NULL
	for (i in 0:6) {
		out1=lsoda(y=y0,times=seq(2*i,2*(i+1),.2), parms=c(lambda1=lambda1,lambda2=0,b=b,d=d,e=eT,clr=clrT), 
				rtol=1e-4, atol= rep(1e-4,3),
				func = "derivsPhil", dllname = "canceR",initfunc = "initPhil" )
		y0=out1[nrow(out1),2:4]+c(0,0,30)
		outs=rbind(outs,out1)
	}
	g$outT=as.data.frame(outs)
	g$dT$EV=g$outT[match(g$dT$days,g$outT$time),"V"]
	SSE=updateSSE(g$dT,SSE)
	
	outs=NULL
	y0<-c(V=V0,K= K0,g=20)
	for (i in 0:13) {
		out1=lsoda(y=y0,times=seq(i,i+1,.2),parms=c(lambda1=lambda1,lambda2=0,b=b,d=d,e=eA,clr=clrA), 
				rtol=1e-4, atol= rep(1e-4,3),
				func = "derivsPhil", dllname = "canceR",initfunc = "initPhil" )
		y0=out1[nrow(out1),2:4]+c(0,0,20)
		outs=rbind(outs,out1)
	}
	g$outA=as.data.frame(outs)
	g$dA$EV=g$outA[match(g$dA$days,g$outA$time),"V"]
	SSE=updateSSE(g$dA,SSE)
	
	y0<-c(V=V0,K= K0,g=20)
	outs=NULL
	for (i in 0:9) {
		out1=lsoda(y=y0,times=seq(i,i+1,.2), parms=c(lambda1=lambda1,lambda2=0,b=b,d=d,e=eE,clr=clrE), 
				rtol=1e-4, atol= rep(1e-4,3),
				func = "derivsPhil", dllname = "canceR",initfunc = "initPhil" )
		y0=out1[nrow(out1),2:4]+c(0,0,20)
		outs=rbind(outs,out1)
	}
	g$outE=as.data.frame(outs)
	g$dE$EV=g$outE[match(g$dE$days,g$outE$time),"V"]
	SSE=updateSSE(g$dE,SSE)
	
	y0<-c(V=V0valid,K= K0,g=4)  # endo low dose not enough to control tumor
	outs=NULL
	for (i in 0:15) {
		out1=lsoda(y=y0,times=seq(i,i+1,.2), parms=c(lambda1=lambda1,lambda2=0,b=b,d=d,e=eE,clr=clrE), 
				rtol=1e-4, atol= rep(1e-4,3),
				func = "derivsPhil", dllname = "canceR",initfunc = "initPhil" )
		y0=out1[nrow(out1),2:4]+c(0,0,4)
		outs=rbind(outs,out1)
	}
	g$outEL=as.data.frame(outs)
	g$dEL$EV=g$outEL[match(g$dEL$days,g$outEL$time),"V"]
	SSE=updateSSE(g$dEL,SSE)
	
	y0<-c(V=V0valid,K= K0,gA=20,gE=20) # angio and endo high dose combined
	outs=NULL
	for (i in 0:24) {
		out1=lsoda(y=y0,times=seq(i,i+1,.2), parms=c(lambda1=lambda1,lambda2=0,b=b,d=d,
#            eA=0.15,clrA=0.38,eE=0.66,clrE=1.7), rtol=1e-4, atol= rep(1e-4,4),dll="philModAE")
						eA=eA,clrA=clrA,eE=eE,clrE=clrE), rtol=1e-4, atol= rep(1e-4,4),
				func = "derivsPhilAE", dllname = "canceR",initfunc = "initPhilAE" )
		y0=out1[nrow(out1),2:5]+c(0,0,20,20)
		outs=rbind(outs,out1)
	}
	g$outAE=as.data.frame(outs)
	g$dAE$EV=g$outAE[match(g$dAE$days,g$outAE$time),"V"]
	g$SSE=updateSSE(g$dAE,SSE)
	
	detach(Pframe)
	g
} ### 



fopt <- function(pars,model) {
	pars=exp(pars)
	model$params[names(pars),"final"]=pars  
	model=simData(model)
	return(model$SSE)   
}

plotVK<-function(g) {
	windows()
	par(mfrow=c(2,3))
	with(g$outC, {plot(time,V,type="l",ylab="Tumor Size (uL)",xlab="Time (days)",main="Control",ylim=c(0,12000));
				lines(time,K,type="l",col="red")} )
	points(g$dC)
	with(g$outT, {plot(time,V,type="l",ylab="Tumor Size (uL)",xlab="Time (days)",main="TNP470 (30mg/kg/2day)",ylim=c(0,2500));
				lines(time,K,type="l",col="red")})
	points(g$dT)
	with(g$outA, {plot(time,V,type="l",ylab="Tumor Size (uL)",xlab="Time (days)",main="Angiostatin (20mg/kg/day)",ylim=c(0,500));
				lines(time,K,type="l",col="red")})
	points(g$dA)
	with(g$outE, {plot(time,V,type="l",ylab="Tumor Size (uL)",xlab="Time (days)",main="Endostatin (20mg/kg/day)",ylim=c(0,400));
				lines(time,K,type="l",col="red")})
	points(g$dE)
	with(g$outEL, {plot(time,V,type="l",ylab="Tumor Size (uL)",xlab="Time (days)",main="Endostatin (4mg/kg/day)",ylim=c(0,5000));
				lines(time,K,type="l",col="red")})
	points(g$dEL)
	with(g$outAE, {plot(time,V,type="l",ylab="Tumor Size (uL)",xlab="Time (days)",main="Angio+Endo (20mg/kg/day)",ylim=c(0,350));
				lines(time,K,type="l",col="red")})
	points(g$dAE)
}


plotg<-function(g) {
	windows()
	par(mfrow=c(2,3))
	plot.new()
	with(g$outT, plot(time,g,type="l",ylab="drug (mg/kg)",xlab="Time (days)",main="TNP470 (30mg/kg/2day)"))
	with(g$outA, plot(time,g,type="l",ylab="drug (mg/kg)",xlab="Time (days)",main="Angiostatin (20mg/kg/day)"))
	with(g$outE, plot(time,g,type="l",ylab="drug (mg/kg)",xlab="Time (days)",main="Endostatin (20mg/kg/day)"))
	with(g$outEL, plot(time,g,type="l",ylab="drug (mg/kg)",xlab="Time (days)",main="Endostatin (4mg/kg/day)"))
	with(g$outAE, {plot(time,gA,type="l",ylab="drug (mg/kg)",xlab="Time (days)",main="Angio+Endo (20mg/kg/day)",ylim=c(0,65));
				lines(time,gE)})
}

